lib/pbench/server/api/resources/datasets_list.py

```diff
@@ -416,9 +409,78 @@ def _get(
             A JSON response containing the paginated query result
         """
         json = params.query
+        auth_id = Auth.get_current_user_id()
 
         # Build a SQLAlchemy Query object expressing all of our constraints
-        query = Database.db_session.query(Dataset).outerjoin(Metadata)
+
+        """SQL notes:
+
+        We want to be able to query across all four metadata key namespaces,
+        and a SELECT WHERE clause can only match within individual rows.
+        That is, if we look for "value1 = 'a' and value2 = 'b'", we won't
+        find a match with a single LEFT JOIN as the matched Metadata is
+        distributed across separate rows in the join table.
+
+        The way we get around this is ugly and inflexible, but I'm unable to
+        find a better way. We actually JOIN against Metadata four separate
+        times; all are constrained by the Dataset foreign key, and each is
+        additionally constrained by the primary Metadata.key value. (And the
+        "user" namespace is additionally constrained by the authorized user
+        ID, and omitted if we're not authenticated.)
+
+        This results in a join table with a JSON value column for each of the
+        tables, so that a single SELECT WHERE can match against all four of the
+        namespaces on a single row in the join table. In order to be able to
+        access the duplicate Metadata.value columns, we first create some
+        SQL name aliases. What we'll end up with, before we start adding our
+        filters, is something (simplified) like:
+
+        Dataset   mtable        stable      gtable          utable
+        --------- ------------- ----------- --------------- --------------
+        drb       {             {           {               {
+                   "pbench":{    "origin":{  "dashboard":{   "dashboard":{
+                   }             }           }               }
+                  }             }           }               }
+        test      {             {           {               {
+                   "pbench":{    "origin":{  "dashboard":{   "dashboard":{
+                   }             }           }               }
+                  }             }           }               }
+        """
+        mtable = aliased(Metadata)
+        stable = aliased(Metadata)
+        gtable = aliased(Metadata)
+        utable = aliased(Metadata)
```

Comment:
Should we name them as `m_table`, `s_table` and so on, felt confusing when I read it first.

Reply:
Yeah, I considered something longer and less wieldy after typing "stable" myself, and decided I'd rather keep them short. In fact, considering that the scope is narrow and they're really referenced through the `dict` after this setup, I'd considered just `m`, `s`, `g`, `u`.

Reply:
The change to iterate over the dictionary finessed this issue very nicely.